src/mechanics/mesh/MeshGmsh.h

```diff
@@ -0,0 +1,93 @@
+#pragma once
+
+#include "MeshFem.h"
+#include "base/Group.h"
+#include <string>
+#include <vector>
+#include <map>
+
+class GmshFileContent;
+
+namespace NuTo
+{
+
+//! @brief Reads a gmsh msh-file and converts it to a MeshFEM. Additionally physical groups are created and can be
+//! accessed via ID or name
+class MeshGmsh
+{
+public:
+    //! @brief ctor
+    //! @remark deleted: The class is always constructed with its associated files name
+    MeshGmsh() = delete;
+
+    //! @brief copy ctor
+    //! @remark deleted: Member MeshFEM is not copyable
+    MeshGmsh(const MeshGmsh&) = delete;
+
+    //! @brief move ctor
+    MeshGmsh(MeshGmsh&&) = default;
+
+    //! @brief copy assignment operator
+    //! @remark deleted: Member MeshFEM is not copyable
+    MeshGmsh& operator=(const MeshGmsh&) = delete;
+
+    //! @brief move assignment operator
+    MeshGmsh& operator=(MeshGmsh&&) = default;
+
+    //! @brief dtor
+    ~MeshGmsh() = default;
+
+    //! @brief ctor
+    //! @param fileName: Name of the file (including path)
+    explicit MeshGmsh(const std::string& fileName);
+
+    //! @brief Gets an element collection group associated to the physical name defined in gmsh
+    //! @param physicalName: Group name (defined in gmsh)
+    //! @return Desired element collection group
+    const NuTo::Group<ElementCollection>& GetPhysicalGroup(std::string physicalName) const;
+
+
+    //! @brief Gets an element collection group associated to the physical ID defined in gmsh
+    //! @param physicalGroupId: gmsh ID
+    //! @return Desired element collection group
+    const NuTo::Group<ElementCollection>& GetPhysicalGroup(int physicalGroupId) const;
+
+private:
+    //! @brief Adds an element to its corresponding physical group. If the element is part of named group and the group
+    //! does not exist, then a named physical group is also created.
+    //! @param fileContent: Special structure (see MeshGmsh.cpp) that holds the read file content
+    //! @param element: Current element
+    //! @param physicalGroupId: Id of the elements physical group
+    void AddElementToPhysicalGroup(const GmshFileContent& fileContent, ElementCollection& element, int physicalGroupId);
```

Comment:
1) I would avoid using return arguments as much as possible. This may not be possible here due to not copyable objects.
2) Please provide the `r` prefix for return arguments. So `rElement`. Same goes for all the other signatures with nonconst references.

Reply:
Will see which kind of `std::tie` and return value magic we can do to change that. If it is not possible I will add the r prefix.